lightning/src/ln/router.rs

```diff
@@ -967,19 +983,42 @@ impl Router {
 			if first_hops.is_none() || hop.src_node_id != network.our_node_id { // first_hop overrules last_hops
 				if network.nodes.get(&hop.src_node_id).is_some() {
 					if first_hops.is_some() {
-						if let Some(first_hop) = first_hop_targets.get(&hop.src_node_id) {
-							add_entry!(first_hop, hop.src_node_id, dummy_directional_info, 0);
+						if let Some(&(ref first_hop, ref features)) = first_hop_targets.get(&hop.src_node_id) {
+							// Currently there are no channel-context features defined, so we are a
+							// bit lazy here. In the future, we should pull them out via our
+							// ChannelManager, but there's no reason to waste the space until we
+							// need them.
+							add_entry!(first_hop, hop.src_node_id, dummy_directional_info, ChannelFeatures::relevant_init_flags_to_channel(&features), 0);
 						}
 					}
-					add_entry!(hop.short_channel_id, target, hop, 0);
+					// BOLT 11 doesn't allow inclusion of features for the last hop hints, which
```

Comment:
Because we may want to use var_onion_optin for routing to last_hop but we don't have any connection or receive yet a node_announcement for him ? Even wonder if not allowing this is bad for privacy because it forces node to advertise its existence on the net or allow incoming connections instead of few selected gateways.

Reply:
Note that there's currently no reason to need TLV-based onion for second-to-last-hops (and the invoice *does* provide features for last-hops), so its no big deal. It should probably be fixed eventually, but it doesn't impact B-AMP or so currently. May be worth opening an issue on the BOLT repo.

Reply:
You mean like amending `r` field to provide features with routing hints ? You can assume than receiving is only to use nodes feature-compliant with its payment scheme, but we may have more complex schemes though

Reply:
Right, ideally r fields could hold feature bits in some form for those nodes, but no reason to need it today.